org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/NodeModelBuilder.java

```diff
@@ -292,6 +293,20 @@ public void replaceAndTransferLookAhead(INode oldNode, INode newRootNode) {
 		}
 	}
 
+	public void replaceAndTransferLookAhead(INode oldNode, INode newRootNode) {
+		AbstractNode newNode = doReplaceAndTransferLookAhead(oldNode, newRootNode);
+		ICompositeNode root = newNode.getRootNode();
+		adjustLookAhead(0, ((AbstractNode) root).basicIterator());
+	}
+
+	/**
+	 * @since 2.26
+	 */
+	 public void replaceAndTransferLookAheadKeepingOffset(INode oldNode, INode newRootNode) {
```

Comment:
I'm afraid this cannot be accepted without a test that proves the following scenario impossible:

Assuming we have a model with only three leaf nodes ( Name(totalOffset, totalEndOffset) ):
- L1 (0, 10)
- L2 (10, 20)
- L3 (20, 30)

`oldNode` encloses L2, `newRootNode` has a single leaf beneath of length 5 ( L4(0, 5) at this point ).
Now we are calling `replaceAndTransferLookAheadKeepingOffset` effectively ripping out `oldNode` with its L2 and replacing that one by the first child of newRootNode with its L4.
According to the logic, L4 would be adjusted towards L4(10, 15). Afterwards, the model is inconsistent since there's a gap of 5 characters before L3 starts at offset 20.

Reply:
I will check this.

Reply:
@rubenporras Thank you very much for upstreaming the fixes you use in production since quite some time. But please keep in mind that we cannot accept this patch without at least some test coverage.

Reply:
Hi @szarnekow ,

yes, I understand that I need to add test coverage, it will take a while for me to find time to understand how to submit a test for it, but I will come to it eventually. Is it fine for you if I keep the PR open for the time being?

Reply:
Sure, that works for me.

Reply:
Not yet, I promised to write some unit tests, but I did not have time for it so far.